File: cli/triada/keyboard.py
"""ES QWERTY (Spain) layout knowledge, shared by the CLI drills.

Mirrors the KB table in web/app.js.
"""
from __future__ import annotations

# rows of (primary, shifted, altgr, finger)
KB = [
    [("º", "ª", "\\", 4), ("1", "!", "|", 4), ("2", '"', "@", 3), ("3", "·", "#", 2),
     ("4", "$", "~", 1), ("5", "%", "€", 1), ("6", "&", "¬", 1), ("7", "/", "", 1),
     ("8", "(", "", 2), ("9", ")", "", 3), ("0", "=", "", 4), ("'", "?", "", 4),
     ("¡", "¿", "", 4)],
    [("q", "Q", "", 4), ("w", "W", "", 3), ("e", "E", "", 2), ("r", "R", "", 1),
     ("t", "T", "", 1), ("y", "Y", "", 1), ("u", "U", "", 1), ("i", "I", "", 2),
     ("o", "O", "", 3), ("p", "P", "", 4), ("`", "^", "[", 4), ("+", "*", "]", 4)],
    [("a", "A", "", 4), ("s", "S", "", 3), ("d", "D", "", 2), ("f", "F", "", 1),
     ("g", "G", "", 1), ("h", "H", "", 1), ("j", "J", "", 1), ("k", "K", "", 2),
     ("l", "L", "", 3), ("ñ", "Ñ", "", 4), ("´", "¨", "{", 4), ("ç", "Ç", "}", 4)],
    [("<", ">", "", 4), ("z", "Z", "", 4), ("x", "X", "", 3), ("c", "C", "", 2),
     ("v", "V", "", 1), ("b", "B", "", 1), ("n", "N", "", 1), ("m", "M", "", 2),
     (",", ";", "", 3), (".", ":", "", 4), ("-", "_", "", 4)],
]

DEAD = {
    "á": ("´", "a"), "é": ("´", "e"), "í": ("´", "i"), "ó": ("´", "o"), "ú": ("´", "u"),
    "Á": ("´", "A"), "É": ("´", "E"), "Í": ("´", "I"), "Ó": ("´", "O"), "Ú": ("´", "U"),
    "ü": ("¨", "u"), "Ü": ("¨", "U"), "à": ("`", "a"), "è": ("`", "e"),
}

HANDS = {}
INDEX = {}
for _r, _row in enumerate(KB):
    for _c, _k in enumerate(_row):
        left = _c < len(_row) // 2
        if _k[0] and _k[0] not in INDEX:
            INDEX[_k[0]] = (_r, _c, "")
        if _k[1] and _k[1] not in INDEX:
            INDEX[_k[1]] = (_r, _c, "Shift")
        if _k[2] and _k[2] not in INDEX:
            INDEX[_k[2]] = (_r, _c, "AltGr")

FINGER_NAME = {1: "index", 2: "middle", 3: "ring", 4: "pinky"}


def key_hint(ch: str) -> str:
    if ch == " ":
        return "space bar — thumb"
    if ch == "\n":
        return "Enter — right pinky"
    if ch in DEAD:
        dead, base = DEAD[ch]
        mod = "Shift + " if INDEX.get(dead, ("", "", ""))[2] == "Shift" else ""
        return f"dead key {mod}{dead}, then {base}"
    pos = INDEX.get(ch)
    if not pos:
        return ""
    r, c, mod = pos
    key = KB[r][c]
    hand = "left" if c < len(KB[r]) // 2 else "right"
    finger = FINGER_NAME[key[3]]
    prefix = f"{mod} + " if mod else ""
    return f"{prefix}{key[0]}   ({hand} {finger})"

File: cli-rs/src/keyboard.rs
//! ES QWERTY (Spain) layout knowledge — port of cli/triada/keyboard.py.

use std::collections::HashMap;
use std::sync::OnceLock;

// (primary, shifted, altgr, finger)
type Key = (&'static str, &'static str, &'static str, u8);

pub const KB: &[&[Key]] = &[
    &[
        ("º", "ª", "\\", 4), ("1", "!", "|", 4), ("2", "\"", "@", 3), ("3", "·", "#", 2),
        ("4", "$", "~", 1), ("5", "%", "€", 1), ("6", "&", "¬", 1), ("7", "/", "", 1),
        ("8", "(", "", 2), ("9", ")", "", 3), ("0", "=", "", 4), ("'", "?", "", 4),
        ("¡", "¿", "", 4),
    ],
    &[
        ("q", "Q", "", 4), ("w", "W", "", 3), ("e", "E", "", 2), ("r", "R", "", 1),
        ("t", "T", "", 1), ("y", "Y", "", 1), ("u", "U", "", 1), ("i", "I", "", 2),
        ("o", "O", "", 3), ("p", "P", "", 4), ("`", "^", "[", 4), ("+", "*", "]", 4),
    ],
    &[
        ("a", "A", "", 4), ("s", "S", "", 3), ("d", "D", "", 2), ("f", "F", "", 1),
        ("g", "G", "", 1), ("h", "H", "", 1), ("j", "J", "", 1), ("k", "K", "", 2),
        ("l", "L", "", 3), ("ñ", "Ñ", "", 4), ("´", "¨", "{", 4), ("ç", "Ç", "}", 4),
    ],
    &[
        ("<", ">", "", 4), ("z", "Z", "", 4), ("x", "X", "", 3), ("c", "C", "", 2),
        ("v", "V", "", 1), ("b", "B", "", 1), ("n", "N", "", 1), ("m", "M", "", 2),
        (",", ";", "", 3), (".", ":", "", 4), ("-", "_", "", 4),
    ],
];

fn dead_keys() -> &'static HashMap<char, (char, char)> {
    static MAP: OnceLock<HashMap<char, (char, char)>> = OnceLock::new();
    MAP.get_or_init(|| {
        let mut m = HashMap::new();
        for (acc, dead, base) in [
            ('á', '´', 'a'), ('é', '´', 'e'), ('í', '´', 'i'), ('ó', '´', 'o'), ('ú', '´', 'u'),
            ('Á', '´', 'A'), ('É', '´', 'E'), ('Í', '´', 'I'), ('Ó', '´', 'O'), ('Ú', '´', 'U'),
            ('ü', '¨', 'u'), ('Ü', '¨', 'U'), ('à', '`', 'a'), ('è', '`', 'e'),
        ] {
            m.insert(acc, (dead, base));
        }
        m
    })
}

#[derive(Clone, Copy)]
struct Loc {
    row: usize,
    col: usize,
    modifier: &'static str, // "" | "Shift" | "AltGr"
}

fn index_map() -> &'static HashMap<char, Loc> {
    static MAP: OnceLock<HashMap<char, Loc>> = OnceLock::new();
    MAP.get_or_init(|| {
        let mut m = HashMap::new();
        for (r, row) in KB.iter().enumerate() {
            for (c, k) in row.iter().enumerate() {
                if !k.0.is_empty() {
                    if let Some(ch) = k.0.chars().next() {
                        m.entry(ch).or_insert(Loc { row: r, col: c, modifier: "" });
                    }
                }
                if !k.1.is_empty() {
                    if let Some(ch) = k.1.chars().next() {
                        m.entry(ch).or_insert(Loc { row: r, col: c, modifier: "Shift" });
                    }
                }
                if !k.2.is_empty() {
                    if let Some(ch) = k.2.chars().next() {
                        m.entry(ch).or_insert(Loc { row: r, col: c, modifier: "AltGr" });
                    }
                }
            }
        }
        m
    })
}

fn finger_name(f: u8) -> &'static str {
    match f {
        1 => "index",
        2 => "middle",
        3 => "ring",
        _ => "pinky",
    }
}

pub fn key_hint(ch: char) -> String {
    if ch == ' ' {
        return "space bar — thumb".to_string();
    }
    if ch == '\n' {
        return "Enter — right pinky".to_string();
    }
    if let Some(&(dead, base)) = dead_keys().get(&ch) {
        let modf = index_map().get(&dead).map(|l| l.modifier).unwrap_or("");
        let mod_prefix = if modf == "Shift" { "Shift + " } else { "" };
        return format!("dead key {}{}, then {}", mod_prefix, dead, base);
    }
    let loc = match index_map().get(&ch) {
        Some(l) => *l,
        None => return String::new(),
    };
    let key = KB[loc.row][loc.col];
    let hand = if loc.col < KB[loc.row].len() / 2 { "left" } else { "right" };
    let finger = finger_name(key.3);
    let prefix = if !loc.modifier.is_empty() { format!("{} + ", loc.modifier) } else { String::new() };
    format!("{}{}   ({} {})", prefix, key.0, hand, finger)
}
